File: internal/ui/cursor.go
package ui

func (e *Editor) Focus() {
	if e.Row > e.Y + e.ROWS { e.Y = e.Row + e.ROWS }
	if e.Row < e.Y { e.Y = e.Row }
}

func (e *Editor) FocusCenter() {
	e.Screen.Show()
	if e.Row > e.Y + e.ROWS {
		e.Y = e.Row + e.ROWS
	}
	if e.Row < e.Y {
		e.Y = e.Row
	}

	e.Y -= e.ROWS/2
	if e.Y < 0 { e.Y = 0 }

	centerRow := e.ROWS / 2
	// Update the cursor row to the center row if necessary
	if e.Row - e.Y > centerRow {
		e.Y += e.Row - e.Y - centerRow
	}
}

func (e *Editor) FocusProcessPanel() {
	e.Screen.Show()
	if e.ProcessPanelCursorY > e.ProcessPanelScroll + e.ROWS {
		e.ProcessPanelScroll = e.ProcessPanelCursorY + e.ROWS
	}
	if e.ProcessPanelCursorY < e.ProcessPanelScroll {
		e.ProcessPanelScroll = e.ProcessPanelCursorY
	}

	e.ProcessPanelScroll -= e.ProcessPanelHeight/2
	if e.ProcessPanelScroll < 0 { e.ProcessPanelScroll = 0 }

	centerRow := e.ProcessPanelHeight / 2
	// Update the cursor row to the center row if necessary
	if e.ProcessPanelCursorY - e.ProcessPanelScroll > centerRow {
		e.ProcessPanelScroll += e.ProcessPanelCursorY - e.ProcessPanelScroll - centerRow
	}
}

File: source/cursor.cpp
/**
    Copyright (C) 2023 - 2026, edgo authors

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, see <https://www.gnu.org/licenses/>.
*/

#include "editor.h"

void Focus(void) {
	if (e.row > e.y + e.ROWS) {
		e.y = e.row + e.ROWS;
	} else if (e.row < e.y) {
		e.y = e.row;
	}
}

void FocusCenter(void) {
//	e.Screen.Show()
	if (e.row > e.y + e.ROWS) {
		e.y = e.row + e.ROWS;
	} else if (e.row < e.y) {
		e.y = e.row;
	}

	e.y -= e.ROWS / 2;
	if (e.y < 0) { e.y = 0; }

	int centerRow = e.ROWS / 2;
	// Update the cursor row to the center row if necessary
	if (e.row - e.y > centerRow) {
		e.y += e.row - e.y - centerRow;
	}
}
